### ISSUE TITLE
Problem with async generators - TypeError: Object.setPrototypeOf called on null or undefined using Babel 6.x.x

### ISSUE THREAD

**Issue by kyrylkov** [babel/babel#2744]
.babelrc:

```
{
  "presets": ["es2015", "stage-3"]
}
```

```
C:\Users\User\Dev\project>nodemon
[nodemon] 1.8.0
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `babel-node server.js`

C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-polyfill\node_modules\regenerator\runtime.js:111
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
             ^

TypeError: Object.setPrototypeOf called on null or undefined
    at Function.setPrototypeOf (native)
    at runtime.mark (C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-polyfill\node_modules\regenerator\runtime.js:111:14)
    at Array.map (native)
    at rest.js:105:37
    at Object.<anonymous> (rest.js:117:2)
    at Module._compile (module.js:435:26)
    at loader (C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-core\lib\api\register\node.js:136:5)
    at Object.require.extensions.(anonymous function) [as .js] (C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-core\lib\api\register\node.js:146:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (routes.js:4:14)
    at Module._compile (module.js:435:26)
    at loader (C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-core\lib\api\register\node.js:136:5)
    at Object.require.extensions.(anonymous function) [as .js] (C:\Users\User\AppData\Roaming\npm\node_modules\babel-cli\node_modules\babel-core\lib\api\register\node.js:146:7)
```

**Comment by johngeorgewright** [comment]
I've got the same problem.

```js
async function test() {
}
```

```sh
$ babel-node test.js
node_modules/babel-polyfill/node_modules/regenerator/runtime.js:111
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
             ^

TypeError: Object.setPrototypeOf called on null or undefined
    at Function.setPrototypeOf (native)
    at runtime.mark (node_modules/babel-polyfill/node_modules/regenerator/runtime.js:111:14)
    at Array.map (native)
    at app.js:1:22
    at Object.<anonymous> (app.js:2:2)
    at Module._compile (module.js:425:26)
    at loader (node_modules/babel-core/lib/api/register/node.js:136:5)
    at Object.require.extensions.(anonymous function) [as .js] (node_modules/babel-core/lib/api/register/node.js:146:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
```

**Comment by ariporad** [comment]
+1. I'm also having this problem (Happy to provide more detail if needed).

**Comment by ZhengzhongZhang** [comment]
+1, I found it in browser version, node_modules/babel-polyfill/dist/polyfill.js, and chrome told me  
>genFun = undefined

**Comment by koistya** [comment]
Is there a workaround for this issue?

**Comment by kyrylkov** [comment]
@koistya  AFAIK no.  Babel team hasn't commented yet.  Babel 6 is still a total mess.

**Comment by atonparker** [comment]
Seeing this too. I used use `babel --optional es7.asyncFunctions` almost everywhere. Still looking for a working equivalent in Babel 6.0.

**Comment by ZhengzhongZhang** [comment]
@kyrylkov maybe it's a different problem. It's about somethine undefined in polyfill.js. #2643 is because they use the wrong plugin, syntax plugin won't transfrom the code but transform plugin deos
@atonparker 
>Babel 6 ships without any default transforms, so when you run Babel on a file it will just print it back out to you without changing anything.
 from [babel blog](http://babeljs.io/blog/2015/10/31/setting-up-babel-6/)

you can read [http://babeljs.io/docs/plugins/](http://babeljs.io/docs/plugins/)

**Comment by kyrylkov** [comment]
@ZhengzhongZhang I read info about plugins.  `stage-3` includes `transform-async-to-generator`.  I have no idea what else should be included.

**Comment by ZhengzhongZhang** [comment]
I think there is something wrong with babel-polyfill, according to the error. Maybe it's  a bug

**Comment by kyrylkov** [comment]
@ZhengzhongZhang definitely looks like a bug to me

**Comment by vtambourine** [comment]
I noticed that after compiling code with `babel/polyfill` imported, actual requiring of polyfill moved at the very end of file. Can that be a problem, or all modules requiring are hoisted?

**Comment by leebenson** [comment]
same error. To re-create:

* node v4.2.1
* babel 6.1.1 (babel-core 6.0.20)

**.babelrc**

```
{
  "presets": [
    "es2015",
    "stage-0"
  ]
}
```

**index.js**

```
async function createPromise() {
  return true;
}

async function doSomething() {
  await createPromise();
  console.log('it worked');
}

doSomething();
```

`babel-node index.js` throws this error:

```
/usr/local/lib/node_modules/babel-cli/node_modules/babel-polyfill/node_modules/regenerator/runtime.js:111
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
             ^
TypeError: Object.setPrototypeOf called on null or undefined
    at Function.setPrototypeOf (native)
    at runtime.mark (/usr/local/lib/node_modules/babel-cli/node_modules/babel-polyfill/node_modules/regenerator/runtime.js:111:14)
    at Array.map (native)
    at index.js:1:29
    at Object.<anonymous> (index.js:3:2)
    at Module._compile (module.js:435:26)
    at loader (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/api/register/node.js:136:5)
    at Object.require.extensions.(anonymous function) [as .js] (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/api/register/node.js:146:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
```

**Comment by breeeew** [comment]
yep, same problem..
node v5.0.0
presets es2015, stage-0

**Comment by atonparker** [comment]
This isn't a fix for this bug, but I was able to get async functions working in node 5 with this .babelrc file:

```
{
  "plugins": [
    "transform-es2015-modules-commonjs",
    "transform-es2015-function-name",
    "transform-async-to-generator"
  ]
}
```

and with `babel-polyfill` installed (but not explicitly required, or you will get an error). It would be nice to use the presets here though.

**Comment by preco21** [comment]
I'm not sure my issue is same this issue.
The error caused after updating the `babel 6.x`.
But similar issue in webpack + babel(with runtime)

webpack.config.js:
```javascript
// ...

module.exports = {
    target: 'web',
    debug: true,
    devtool: '#source-map',
    entry: config.webpack.entry,
    output: {
        path: `${__dirname}/${config.webpack.output}`,
        filename: config.webpack.dest
    },
    module: {
        loaders: [{
            test: /\.js$/,
            exclude: /node_modules/,
            loader: 'babel'
        }],
        postLoaders: [{
            test: /\.json$/,
            exclude: /node_modules/,
            loader: 'json'
        }]
    },
    plugins: [
        new webpack.optimize.CommonsChunkPlugin(config.webpack.commons)
    ]
};
```

.baberc:
```
{
  "presets": [
    "es2015",
    "stage-0"
  ],
  "plugins": [
    "transform-runtime"
  ]
}
```

When I run my code, browser console returns this error:
![image 002](https://cloud.githubusercontent.com/assets/4936201/10956186/eb584ac4-839c-11e5-991a-81be4a28f64a.png)

![image 004](https://cloud.githubusercontent.com/assets/4936201/10956205/12457026-839d-11e5-8e32-64e313045204.png)

* node v5.0.0
* babel v6.1.1 (babel-core 6.0.20)
* webpack v1.12.2

**Comment by reggi** [comment]
I'm looking at this issue here https://github.com/babel/babel/issues/2857

I don't understand why the dependency for `regeneratorRuntime` isn't automatically added to the build, and why I have to require / import `babel-polyfill` to get it to work.

**Comment by Witcher42** [comment]
I have similar problem, after some tracing I found that *transform-regenerator* doesn't work properly. It will throw an Exception after finding a function that doesn't have an id (function name). But other plugins generated a anonymous generator function before *transform-regenerator*. That anonymous generator function doesn't have a function name and that's why I get an error.

Even though I have no idea about why it throws a exception, I found a new [commit](https://github.com/babel/babel/commit/940b86dadbd0151c33c02e89f0b5ff61077c9214#diff-309ce65a0610771fbf1a2e6c2b11bfac) is pulled to *babel-plugin-transform-regenerator* and the patch fixes this problem perfectly. Thanks to @hzoo .

It seems that this fix will be released to npm in the next version?
